custom_components/hafo/sensor.py

```diff
@@ -0,0 +1,134 @@
+"""Sensor platform for Home Assistant Forecaster."""
+
+import logging
+from typing import Any
+
+from homeassistant.components.sensor import SensorEntity, SensorStateClass
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+from homeassistant.util import dt as dt_util
+
+from .const import ATTR_FORECAST, ATTR_HISTORY_DAYS, ATTR_LAST_UPDATED, ATTR_SOURCE_ENTITY, DOMAIN
+from .coordinator import HafoDataUpdateCoordinator
+from .forecasters.historical_shift import ForecastResult
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,  # noqa: ARG001
+    entry: ConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up HAFO sensors based on a config entry."""
+    coordinator: HafoDataUpdateCoordinator = entry.runtime_data
+
+    entities: list[SensorEntity] = [
+        HafoForecastSensor(coordinator),
+    ]
+
+    async_add_entities(entities)
+
+
+class HafoForecastSensor(CoordinatorEntity[HafoDataUpdateCoordinator], SensorEntity):
+    """Sensor that provides forecast data from historical statistics."""
+
+    _attr_has_entity_name = True
+    _attr_state_class = SensorStateClass.MEASUREMENT
+
+    def __init__(self, coordinator: HafoDataUpdateCoordinator) -> None:
+        """Initialize the forecast sensor.
+
+        Args:
+            coordinator: The data update coordinator
+
+        """
+        super().__init__(coordinator)
+
+        # Set up entity attributes
+        self._attr_unique_id = f"{coordinator.entry.entry_id}_forecast"
+        self._attr_name = "Forecast"
+
+        # Copy unit of measurement from source entity if available
+        self._source_entity = coordinator.source_entity
+        self._update_from_source_entity()
+
+    def _update_from_source_entity(self) -> None:
+        """Update sensor attributes from the source entity."""
+        source_state = self.coordinator.hass.states.get(self._source_entity)
```

Comment:
This information should come from the coordinators as in some of the cases there could be multiple source or destination entities, so just pass this through in the coordinator data and update it in the update from data coordinator code.